src/nbl/video/CVulkanPhysicalDevice.h

```diff
@@ -1240,149 +1284,814 @@ class CVulkanPhysicalDevice final : public IPhysicalDevice
 
         return m_bufferUsages[format];
     }
-            
+ 
 protected:
-    core::smart_refctd_ptr<ILogicalDevice> createLogicalDevice_impl(const ILogicalDevice::SCreationParams& params) override
+
+    core::smart_refctd_ptr<ILogicalDevice> createLogicalDevice_impl(ILogicalDevice::SCreationParams&& params) override
     {
-        auto insertFeatureIfAvailable = [this](const ILogicalDevice::E_FEATURE feature, auto& featureSet) -> bool
+        core::unordered_set<core::string> extensionsToEnable;
+
+        // We might alter it to account for dependancies.
+        SFeatures& enabledFeatures = params.featuresToEnable;
+
+        VkPhysicalDeviceVulkan11Features vulkan11Features = { VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VULKAN_1_1_FEATURES, nullptr };
+
+        VkBaseInStructure * lastFeatureInChain = reinterpret_cast<VkBaseInStructure*>(&vulkan11Features);
+        VkBaseInStructure * firstFeatureInChain = lastFeatureInChain;
+        // Vulkan has problems with having features in the feature chain that have all values set to false.
+        // For example having an empty "RayTracingPipelineFeaturesKHR" in the chain will lead to validation errors for RayQueryONLY applications.
+        auto addFeatureToChain = [&firstFeatureInChain,&lastFeatureInChain](void* feature) -> void
+        {
+            VkBaseInStructure* toAdd = reinterpret_cast<VkBaseInStructure*>(feature);
+            
+            // For protecting against duplication of feature structures that may be requested to add to chain twice due to extension requirements
+            const bool alreadyAdded = (toAdd->pNext != nullptr || toAdd == lastFeatureInChain);
+            // More robust version, traverse chain:
+            //const bool alreadyAdded = [&toAdd]() -> bool 
+            //{
+            //    VkBaseInStructure* current = firstFeatureInChain;
+            //    while(current != nullptr)
+            //    {
+            //        if(current->sType == toAdd->sType)
+            //            return false;
+            //        current = current->pNext;
+            //    }
+            //    return true;
+            //}();
+
+            if(!alreadyAdded)
+            {
+                toAdd->pNext = firstFeatureInChain;
+                firstFeatureInChain = toAdd;
```

Comment:
wait what? why are you pushing to the top and not adding to the end?

Reply:
Yea we can do that too xD.

It's like this because I initially wrote this code where I only stored the head pointer and that's the pointer that goes into Vulkan's pNext.

Then HEAD should be const.